[Card] Expand on Click

hugo-agbonon opened the pull request:
This is an implementation proposal for #1354 : To be able to expand card by clicking on an element such as CardHeader, rather than only being able to do it using the CardExpandable button.

This adds a new `expander` prop for Card child components, which triggers the expand on click.

I also updated the card documentation as to provide a demo of the proposed use.

hugo-agbonon commented:
Just did a fixup on the first commit, if anyone tried the previous version I had forgotten to commit modifications and it didn't work, sorry about that.

shaurya947 commented:
@hugo-agbonon I assume that the `expander` prop is used to determine whether or not the card can be expanded by clicking this child. Correct me if I'm wrong. If that's the case, maybe change the prop name to something more intuitive? Like `canActAsExpander`?

Otherwise, this looks fine to me from a high-level (yet to run the code though). @droberts84 @MichaelTaylor3D you guys have any feedback on this?

oliviertassinari commented on src/card/card-expandable.jsx:
```diff
@@ -100,7 +93,7 @@ const CardExpandable = React.createClass({
       <IconButton
         style={mergedStyles}
-        onClick={this._onExpanding}>
+        onClick={this.props.onExpanding}>
```
Why do we have a onClick and not a onTouchTap?

oliviertassinari commented on src/card/card.jsx:
```diff
@@ -31,6 +32,10 @@ const Card = React.createClass({
       if (this.state.expanded === false && currentChild.props.expandable === true)
         return;
+      if (currentChild.props.expander === true) {
+        currentChild.props.onClick = this._onExpandable;
```
Same question here for the `onClick`

hugo-agbonon commented:
I replaced the onClick events by onTouchTap events, and renamed `expander` to `actAsExpander`

@shaurya947 That's right, any direct child component of `Card` with the prop `actAsExpander` set to true will cause the card to be expanded on a click of this child.

shaurya947 commented:
Sounds good @hugo-agbonon! Thanks!

shaurya947 commented on src/card/card.jsx:
```diff
@@ -31,6 +32,10 @@ const Card = React.createClass({
       if (this.state.expanded === false && currentChild.props.expandable === true)
         return;
+      if (currentChild.props.actAsExpander === true) {
+        currentChild.props.onTouchTap = this._onExpandable;
```
@hugo-agbonon this will not work with React 0.14. Props can no longer be mutated.. see here: https://facebook.github.io/react/blog/2015/10/07/react-v0.14.html#breaking-changes

I'll try to fix this.

hugo-agbonon replied:
@shaurya947 @cgestes already wrote a fix, see #1868 / #1872,

shaurya947 replied:
My bad, missed that. Merged now.
